### DISCUSSION TITLE
Where to put shared code between actions and sensors

### DISCUSSION THREAD

**Discussion by minsis** [StackStorm/st2#5461] [Packs] [answered]
I'm working  on writing a pack to interact  with MS Graph and there's a lot of complexity to account connection. I have what I need for the actions in `actions/lib/` but [the docs](https://docs.stackstorm.com/reference/packs.html#anatomy-of-a-pack)  show  that `sensors/common/` is where you would place that module. Is there a more higher level directory that I should be putting my modules  in that both actions and sensors could access?

I'm assuming I could just put `common/` top level, but I need to make sure that should be okay since its never defined in the docs that way, and since most packs don't seem to implement `__init__.py` in any of the directories I'm thinking st2 doesn't do the traditional way of importing python modules from packs.

**Comment by minsis** [comment, marked as answer]
Well searching for another answer in the docs I found this https://docs.stackstorm.com/reference/sharing_code_sensors_actions.html. So I guess I have my answer!

**Reply by ghost** [reply]
@minsis  After operation according to the docs, it did not take effect. May I ask why I missed the configuration？

`from squared import webSocketListen`in sensor is not valid

![image](https://user-images.githubusercontent.com/49810599/146215158-41013164-9374-4fbf-9e75-661ce7005e3c.png)

**Reply by minsis** [reply]
Can you show the import statements from the action you're trying to run?

**Reply by minsis** [reply]
It would also be helpful to have any error output.

**Reply by ghost** [reply]
1. 
![image](https://user-images.githubusercontent.com/49810599/146289951-f5694f3f-ba9b-4905-ab38-a404cda76c5c.png)
2.
![image](https://user-images.githubusercontent.com/49810599/146290068-df54a028-34a5-45c2-9149-f370a351d7ad.png)

**Reply by ghost** [reply]
@minsis  By the way, why can't st2reactor be found in line 8 of Figure 2？

**Reply by minsis** [reply]
Did you even try to execute any of this code in StackStorm? What you're showing me is IDE reference errors, most likely because you don't have a proper Python environment setup in PyCharm. You're asking a lot of support questions that have nothing to do with StackStorm and all to do with your environment. You should probably learn and understand how Jetbrains and Python works. There are plenty of tutorials and videos out there.

For Jetbrains: https://intellij-support.jetbrains.com/hc/en-us/community/topics
For Python: Corey Shafer has very good tutorial videos (best I've seen personally) https://www.youtube.com/channel/UCCezIgC97PvUuR4_gbFUs5g

If you have the professional (paid) version of Pycham you can reach Jetbrains support directly.

**Reply by ghost** [reply]
@minsis 
Thank you very much for your reminder, 11 lines are indeed an IDE env problem, but the line 8 `st2Reactor` seems to have nothing to do with the IDE env.

I found st2client in /usr/local/lib/python3.9/site-packages/ directory, did not find st2Reactor.
![image](https://user-images.githubusercontent.com/49810599/146295126-f2463daf-3a2c-4be4-a67d-f0bc6d7c08a4.png)

![image](https://user-images.githubusercontent.com/49810599/146295574-dabc19fd-bb61-4c17-b077-e986f544f123.png)

**Reply by minsis** [reply]
You can install it by downloading the st2 repo: https://github.com/StackStorm/st2/tree/master/st2reactor

```shell
git clone https://github.com/StackStorm/st2.git
cd st2/st2reactor
pip install .
```

I think the only published pypi package by StackStorm dev is st2client. The rest you have to install yourself.

See: https://docs.stackstorm.com/development/index.html#setting-up-a-development-environment

**Reply by minsis** [reply]
https://docs.stackstorm.com/development/sources.html#project-requirements

**Reply by ghost** [reply]
Thank you for your patience.
